lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -2990,9 +2990,36 @@ PotentialArchetype *PotentialArchetype::updateNestedTypeForConformance(
 void ArchetypeType::resolveNestedType(
                                     std::pair<Identifier, Type> &nested) const {
   auto genericEnv = getGenericEnvironment();
+  Type interfaceType = getInterfaceType();
+
+  // Handle opened existentials.
+  if (auto exist = getOpenedExistentialType()) {
+    if (auto *proto = exist->getAs<ProtocolType>()) {
+      auto protoDecl = proto->getDecl();
+      genericEnv = protoDecl->getGenericEnvironment();
+      interfaceType = protoDecl->getProtocolSelfType();
+    } else {
+      // For a composition type, construct a compound generic signature
+      // and create a new generic environment out of it.
+      auto builder = GenericSignatureBuilder(exist->getASTContext());
+
+      auto layout = exist->getExistentialLayout();
+      auto firstProto = layout.getProtocols().front()->getDecl();
+
+      builder.addGenericParameter(firstProto->getProtocolSelfType());
+      interfaceType = firstProto->getProtocolSelfType();
+
+      for (auto *proto: layout.getProtocols()) {
+        for (auto &req : proto->getDecl()->getRequirementSignature())
+          builder.addRequirement(req, FloatingRequirementSource::forAbstract(),
+                                 nullptr);
+      }
+      auto newSig = std::move(builder).computeGenericSignature(SourceLoc());
+      genericEnv = newSig->createGenericEnvironment();
```

Comment:
This looks inefficient since you're creating a lot of generic environments. They live forever and do not get deallocated. Since the constrained type is fully concrete, it must have been made fully concrete in one of the inherited protocols, so it should be possible to find that protocol without building a whole new GSB here.

Reply:
Turns out there's an obvious problem with this: conflicts like `P1(A == Int) & P2(A == Bool)` will not be catched. Protocol compositions don't have generic environments right? Perhaps there's a more efficient means of creating a generic environment or GSB for a composition by choosing a `Self` archetype and adding a conformance requirement for each protocol?